File: codes/Beginners Problems/LightOJ 1006 - Hex-a-bonacci.java
/**
 * Author    : sjsakib.bd
 * Lang      : JAVA
 * Date      : 2015-04-12 00:26:12
 * Problem   : 1006 - Hex-a-bonacci
 * CPU       : 0.340
 * Memory    : 27276
**/
import java.util.Scanner;
class Main {
	public static void main(String[] args) {
		Scanner s  = new Scanner(System.in);
		int n, caseno = 0, cases;
		cases = s.nextInt();
		while( cases>0 ) {
			int serial[];
			serial = new int[10001];
			for(int i = 0;i<6;i++) {
				serial[i] = s.nextInt() % 10000007;
			}
			n = s.nextInt();
			for(int i = 6;i<=n;i++) {
				serial[i] = (serial[i-1]+serial[i-2]+serial[i-3]+serial[i-4]+serial[i-5]+serial[i-6]) % 10000007;
			}
			System.out.printf("Case %d: %d\n", ++caseno, serial[n]);
			cases--;
		}
	}
}

File: codes/Beginners Problems/LightOJ 1006 - Hex-a-bonacci.cpp
/**
 * Author    : sjsakib.bd
 * Lang      : C++
 * Date      : 2016-09-04 00:51:01
 * Problem   : 1006 - Hex-a-bonacci
 * CPU       : 0.016
 * Memory    : 1916
**/
#include <bits/stdc++.h>

using namespace std;

int mem[10005];
int a, b, c, d, e, f;
int fn( int n ) {
    if( n == 0 )
    {
        mem[n] = a % 10000007;
        return mem[n];
    }
    if( n == 1 )
    {
        mem[n] = b % 10000007;
        return mem[n];
    }
    if( n == 2 )
    {
       mem[n] = c % 10000007;
       return mem[n];
    }
    if( n == 3 )
    {
        mem[n] = d % 10000007;
        return mem[n];
    }
    if( n == 4 )
    {
        mem[n] = e % 10000007;
        return mem[n];
    }
    if( n == 5 )
    {
        mem[n] = f % 10000007;
        return mem[n];
    }
    if(mem[n] != -1)
    {
        return mem[n];
    }
    if(mem[n] == -1)
    {
        mem[n] = (fn(n-1) + fn(n-2) + fn(n-3) + fn(n-4) + fn(n-5) + fn(n-6) )% 10000007 ;
        return mem[n];
    }
}
int main() {
    int n, caseno = 0, cases;
    memset(mem, -1, sizeof(mem));
    scanf("%d", &cases);
    while( cases-- ) {
        scanf("%d %d %d %d %d %d %d", &a, &b, &c, &d, &e, &f, &n);
        printf("Case %d: %d\n", ++caseno, fn(n) % 10000007);
        memset(mem, -1, sizeof(mem));
    }
    return 0;
}